### PR TITLE
feat: pca algorithm

### PR DESCRIPTION
**Pull request by davidbp**
This PR adds PCA class in math.dimensionality_reduction

Some benchmarking the PCA fit method  with  `n=500_000`:

```
n_feat      time (sec)
128        0.6
256        1
512        2
1024     4.5
```

An example to test

```
import numpy as np
import matplotlib.pyplot as plt
from sklearn.datasets import make_blobs

X, y = make_blobs(n_samples=1000, n_features=3,
                  centers = np.array([[0,0,0,0,10],[1,0,0,0,0],[1,1,1,1,1]])
                  )

# plot 2 random coordinates from the original data
plt.scatter(X[:,0], X[:,2], alpha=0.2, c=y)
plt.savefig('./random_features.png')

# plot using DocumentArray.visualize
from jina import Document, DocumentArray
darray = DocumentArray([Document(embedding=x, tags={'houses': str(y_k)}) for x, y_k in zip(X,y)])
darray.visualize(tag_name='houses',file_path='./test.png')
```

Should generate a plot where 2 random components are plotted

random_features

The result of .visualize

test

### CODE REVIEW

File: jina/math/dimensionality_reduction.py [hunk 1]
```diff
@@ -0,0 +1,56 @@
...
+
+        # Sort eigenvalues by magnitude (higher to lower)
+        idx = e_values.argsort()[::-1]
+        e_values = e_values[idx]
+
+        # Sort eigen vectors by eigenvalue magnitude (higher to lower)
+        e_vectors = e_vectors[:, idx]
+
+        # Projection matrix contains eigenvectors sorted
+        self.w = e_vectors
```
Comment (JoanFM) [inline jina/math/dimensionality_reduction.py:34]: Initialize these attributes at initialization time

Comment (davidbp) [reply jina/math/dimensionality_reduction.py:34]: This is not known at initialization time, why set it up?

Comment (JoanFM) [reply jina/math/dimensionality_reduction.py:34]: set it to None perhaps, but initialize it.

Comment (JoanFM) [reply jina/math/dimensionality_reduction.py:34]: and in transform u can check if self w is not None.

Otherwise just using transform will not work

Comment (davidbp) [reply jina/math/dimensionality_reduction.py:34]: done

File: jina/math/dimensionality_reduction.py [hunk 2]
```diff
@@ -0,0 +1,58 @@
+import numpy as np
+
+
+class PCA:
+    """A class for PCA dimensionality reduction in Jina."""
+
+    def __init__(self, n_components: int, whitening: bool = False):
+        self.n_components = n_components
+        self.whitening = whitening
```
Comment (bwanglzu) [inline jina/math/dimensionality_reduction.py:9]: if my memery is correct, whitening should be always true?

File: jina/math/dimensionality_reduction.py [hunk 3]
```diff
@@ -0,0 +1,58 @@
...
+        self.e_values = None
+        self.w = None
```
Comment (bwanglzu) [inline jina/math/dimensionality_reduction.py:11]: what is `w`? `coefficient vector`?

File: jina/types/arrays/neural_ops.py [hunk 4]
```diff
@@ -82,3 +83,26 @@ def match(
...
+    ):
+        """Visualize embeddings in a 2D projection with the PCA algorithm.
+
+        :param colors: Optional np.ndarray of lengh equal to len(self) containing integers for colors.
+        :param file_path: Optional path to store the visualization.
+        """
+
+        import matplotlib.pyplot as plt
+
+        pca = PCA(n_components=2)
```
Comment (bwanglzu) [inline jina/types/arrays/neural_ops.py:98]: `n_components` should be the variable you initialized in constructor?

**Review by bwanglzu — COMMENTED** [review]
left some comments

File: jina/math/dimensionality_reduction.py (hunk 2 above)
Comment (davidbp) [reply jina/math/dimensionality_reduction.py:9]: sklearn does not do it by default, I just made an implementation that is equivalent to the default sklearn version

File: jina/math/dimensionality_reduction.py (hunk 3 above)
Comment (davidbp) [reply jina/math/dimensionality_reduction.py:11]: These are the weights learned by PCA (eigenvectors as columns)

File: jina/types/arrays/neural_ops.py (hunk 4 above)
Comment (davidbp) [reply jina/types/arrays/neural_ops.py:98]: which constructor? It is in the PCA constructor, not in method.visualize

File: tests/unit/types/arrays/test_neural_ops.py [hunk 5]
```diff
@@ -256,3 +258,22 @@ def dotp(x, y):
     for d in D1:
         for m in d.matches:
             assert 'dotp' in m.scores
+
+
+def test_pca_projection(embeddings):
+    n_components = 2
+    n_features = embeddings.shape[1]
+    pca = PCA(n_components=n_components)
```
Comment (tadejsv) [inline tests/unit/types/arrays/test_neural_ops.py:266]: You can also check this with whithening enabled (maybe just parametrize this test)

Comment (davidbp) [reply tests/unit/types/arrays/test_neural_ops.py]: Yes, added as parameter

File: jina/math/dimensionality_reduction.py [hunk 6]
```diff
@@ -0,0 +1,63 @@
...
+
+
+class PCA:
+    """A class for PCA dimensionality reduction in Jina."""
+
+    def __init__(self, n_components: int, whiten: bool = False):
+        """
+
+        :param n_components: Number of components to keep when projecting with PCA
+        :param whiten: Flag variable stating if there projecting with whitening
```
Comment (bwanglzu) [inline jina/math/dimensionality_reduction.py:11]: you can remove docstring here and move to class docstrings, checkout how other classes write docstrings

**Review by bwanglzu — APPROVED** [review]
LGTM!

File: jina/math/dimensionality_reduction.py (hunk 6 above)
Comment (davidbp) [reply jina/math/dimensionality_reduction.py:11]: I will do it now

File: extra-requirements.txt [hunk 7]
```diff
@@ -65,3 +65,4 @@ aiostream:                  devel, cicd, async, client, standard, daemon
 jsonschema:                 cicd
 dgl==0.6.1:                 cicd, graph
 black==20.8b1:              cicd
+matplotlib>=3.3.1:          devel, test
```
Comment (hanxiao) [inline extra-requirements.txt:68]: cicd

**Review by tadejsv — COMMENTED** [review]
Are we going to implement the continous updating in this PR? If not I would approve and merge this

**Review by hanxiao — APPROVED** [review]

**Review by bwanglzu — DISMISSED** [review]
LGTM!

**Review by tadejsv — APPROVED** [review]
LGTM!

**Comment by hanxiao** [comment]
please fix the commit message

**Comment by bwanglzu** [comment]
commit message fixed

File: jina/types/arrays/neural_ops.py [hunk 8]
```diff
@@ -82,3 +83,26 @@ def match(
...
+        x_mat = np.stack(self.get_attributes('embedding'))
+        x_mat_2d = pca.fit_transform(x_mat)
+
+        plt.scatter(x_mat_2d[:, 0], x_mat_2d[:, 1], c=colors)
+        plt.title('PCA projection')
+        plt.xlabel('PCA component 0')
+        plt.ylabel('PCA component 1')
```
Comment (hanxiao) [inline jina/types/arrays/neural_ops.py:105]: - square figure with equal axis, best fit
- please use more informative title,  e.g. '{number of data points} with PCA'
- remove x & y label, not necessary

File: jina/types/arrays/neural_ops.py [hunk 9]
```diff
@@ -82,3 +83,26 @@ def match(
...
+        :param file_path: Optional path to store the visualization.
+        """
+
+        import matplotlib.pyplot as plt
+
+        pca = PCA(n_components=2)
+        x_mat = np.stack(self.get_attributes('embedding'))
+        x_mat_2d = pca.fit_transform(x_mat)
+
+        plt.scatter(x_mat_2d[:, 0], x_mat_2d[:, 1], c=colors)
```
Comment (hanxiao) [inline jina/types/arrays/neural_ops.py:102]: i think it would be useful to make colors a mapping of `doc.tags['some_tag']`, e.g.

basically implement logic to enable the following mapping
```
if tags['dummy_label'] == 'dummy_value': color = 1
if tags['dummy_label'] == 'dummy_value2': color = 2
```

any opinions?

Comment (hanxiao) [reply jina/types/arrays/neural_ops.py:102]: use dot with alpha value gives better visualization

**Review by hanxiao — CHANGES_REQUESTED** [review]
please fix accordingly

File: jina/types/arrays/neural_ops.py (hunk 9 above)
Comment (davidbp) [reply jina/types/arrays/neural_ops.py]: I've updated the code with what you propose, coloring by tag makes sense

File: jina/types/arrays/neural_ops.py (hunk 8 above)
Comment (davidbp) [reply jina/types/arrays/neural_ops.py]: Done
